# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1382027548 14400
#      Thu Oct 17 12:32:28 2013 -0400
# Node ID 9f6e4e5c2bac5b7933be27415086bb732ffb1fdb
# Parent  36917b3db6b2654ad6f824143cd9a2a383c9510f
avoid memory leak in parser (bug #40231)

* oct-parse.in.yy (octave_base_parser::make_try_command): If an
an error identifier is extracted from the list of cleanup statements,
delete the first statement from the list.

diff --git a/libinterp/parse-tree/oct-parse.in.yy b/libinterp/parse-tree/oct-parse.in.yy
--- a/libinterp/parse-tree/oct-parse.in.yy
+++ b/libinterp/parse-tree/oct-parse.in.yy
@@ -2385,16 +2385,19 @@ octave_base_parser::make_try_command (to
             {
               tree_expression *expr = stmt->expression ();
 
               if (expr && expr->is_identifier ())
                 {
                   id = dynamic_cast<tree_identifier *> (expr);
 
                   cleanup_stmts->pop_front ();
+
+                  stmt->set_expression (0);
+                  delete stmt;
                 }
             }
         }
 
       retval = new tree_try_catch_command (body, cleanup_stmts, id,
                                            lc, mc, tc, l, c);
     }
   else
